test/src/search/search_navigation_test.dart

```diff
@@ -0,0 +1,252 @@
+
+    expect(navigationController.value.currentMatchIndex, 1);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 1, extentOffset: 2),
+    );
+
+    navigationController.moveNext();
+
+    expect(navigationController.value.currentMatchIndex, 2);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 2, extentOffset: 3),
+    );
+
+    navigationController.moveNext();
+
+    expect(navigationController.value.currentMatchIndex, 3);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 3, extentOffset: 4),
+    );
+
+    navigationController.moveNext();
+
+    expect(navigationController.value.currentMatchIndex, 0);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 0, extentOffset: 1),
+    );
+  });
+
+  testWidgets('movePrevious()', (wt) async {
+    const text = 'Aaaa';
+    final codeController = await pumpController(wt, text);
+    codeController.selection = const TextSelection.collapsed(offset: 0);
+
+    codeController.showSearch();
+
+    final navigationController =
+        codeController.searchController.navigationController;
+
+    codeController.searchController.settingsController.value =
+        const SearchSettings(
+      isCaseSensitive: false,
+      isRegExp: false,
+      pattern: 'a',
+    );
+
+    expect(navigationController.value.currentMatchIndex, 0);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 0, extentOffset: 1),
+    );
+
+    navigationController.movePrevious();
+
+    expect(navigationController.value.currentMatchIndex, 3);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 3, extentOffset: 4),
+    );
+
+    navigationController.movePrevious();
+
+    expect(navigationController.value.currentMatchIndex, 2);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 2, extentOffset: 3),
+    );
+
+    navigationController.movePrevious();
+
+    expect(navigationController.value.currentMatchIndex, 1);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 1, extentOffset: 2),
+    );
+
+    navigationController.movePrevious();
+
+    expect(navigationController.value.currentMatchIndex, 0);
+    expect(navigationController.value.totalMatchCount, 4);
+    expect(
+      codeController.selection,
+      const TextSelection(baseOffset: 0, extentOffset: 1),
+    );
+  });
+}
+
```

Comment:
Also test setting the selection manually (no change in navigation), then typing text (current match should become `null`).